### ISSUE TITLE
Associations are saved twice, resulting in uniqueness violation

### ISSUE THREAD

**Issue by adeeb1** [rails/rails#40459]
This problem is the same as https://github.com/rails/rails/issues/39173 but with a different setup.

### Steps to reproduce

1. Create two models
2. Create a join table with a `has_many :through` association connecting them
3. Set a uniqueness constraint on the two foreign keys on the join table
4. Define a virtual attribute on one model with an `after_create` callback that assigns the `_ids` (ex: `comment_ids`) value for the other model.

The associations are saved twice, resulting in a uniqueness violation.

```
class Post < ApplicationRecord
  attr_accessor :t_comment_ids

  after_create :associate_with_comments

  has_many :post_comments
  has_many :comments, through: :post_comments

  def associate_with_comments
    self.comment_ids = self.t_comment_ids
  end
end

class Comment < ApplicationRecord
end

class PostComment < ApplicationRecord
  belongs_to :post
  belongs_to :comment
end

comment = Comment.create!
post = Post.create!(t_comment_ids: [comment.id])

# Result:

(0.5ms)  BEGIN
  Post Create (0.9ms)  INSERT INTO "posts" ("created_at", "updated_at") VALUES ($1, $2) RETURNING "id"  [["created_at", "2020-10-27 03:03:20.128366"], ["updated_at", "2020-10-27 03:03:20.128366"]]
  Comment Load (0.6ms)  SELECT "comments".* FROM "comments" WHERE "comments"."id" = $1  [["id", 7]]
  Comment Load (0.6ms)  SELECT "comments".* FROM "comments" INNER JOIN "post_comments" ON "comments"."id" = "post_comments"."comment_id" WHERE "post_comments"."post_id" = $1  [["post_id", 21]]
  PostComment Create (0.5ms)  INSERT INTO "post_comments" ("post_id", "comment_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["post_id", 21], ["comment_id", 7], ["created_at", "2020-10-27 03:03:20.136249"], ["updated_at", "2020-10-27 03:03:20.136249"]]
  PostComment Create (1.3ms)  INSERT INTO "post_comments" ("post_id", "comment_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["post_id", 21], ["comment_id", 7], ["created_at", "2020-10-27 03:03:20.139144"], ["updated_at", "2020-10-27 03:03:20.139144"]]
   (0.3ms)  ROLLBACK
Traceback (most recent call last):
        1: from (irb):8
ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_post_comments_on_post_id_and_comment_id")
DETAIL:  Key (post_id, comment_id)=(21, 7) already exists.
```

I've created a repo at the following link that you can clone with simple instructions in the README: https://github.com/adeeb1/rails_6_1_0_alpha_test

The one-file executable test cases wouldn't work properly with PostgreSQL.

### Expected behavior
A new Post record is saved and associated with the correct Comment records.

### Actual behavior
An exception is thrown: 

```
ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_post_comments_on_post_id_and_comment_id")
```

### System configuration
**Rails version**: 6.1.0-alpha

**Ruby version**: 2.5.7

**Comment by eugeneius** [comment]
You mentioned in https://github.com/rails/rails/issues/39173#issuecomment-716914189 that you ran into this problem "after upgrading to Rails 6.1.0-alpha", but if I downgrade the test application to 6.0.3.4 it fails in the same way. If you intended to report a regression here, there's something missing.

**Comment by adeeb1** [comment]
@eugeneius Thanks for sharing. I will find the latest release that passes the test case. I have an application running 6.0.3.3 in production without this issue, but the test case fails with that version for some odd reason.

**Comment by adeeb1** [comment]
**TL;DR: Use `after_commit` with `on: :create` instead of `after_create`**

I found this issue on all Rails 6 versions, even going back to 5.2. I'm not sure how this is working on my application in production. I found a workaround: use an `after_commit` callback, as suggested in [this comment](https://github.com/rails/rails/issues/39173#issuecomment-633631979). Changing the Post model to the following passes all tests.

```
class Post < ApplicationRecord
  attr_accessor :t_comment_ids

  after_commit :associate_with_comments, on: :create

  has_many :post_comments
  has_many :comments, through: :post_comments

  def associate_with_comments
    self.comment_ids = self.t_comment_ids
  end
end
```

For my needs, this accomplishes exactly the same thing without any drawbacks.

**Comment by adeeb1** [comment]
**NOTE**
I found out that the above is not actually a solution. It causes the virtual attributes to be cleared in some cases. Instead, I settled with setting new join table objects and letting the call to `#save` save the child records. Here's how I changed the model:

```
class Post < ApplicationRecord
  attr_accessor :t_comment_ids

  after_create :associate_with_comments

  has_many :post_comments
  has_many :comments, through: :post_comments

  def associate_with_comments
    new_post_comments = t_comment_ids.map do |new_comment_id|
      self.post_comments.new(comment_id: new_comment_id)
    end
  
    self.post_comments = new_post_comments
  end
end
```

**Comment by stv8** [comment]
I know this is  super old, but I stumbled on this on a fresh rails 8 app. 

I have a User has many roles through user_roles setup


```rb
class User < ApplicationRecord
  # setting the after_create here causes a unique constraint violation
  # if we set it below the relation definition then things work fine 
  after_create :assign_default_role # <----- This errors


  has_many :user_roles
  has_many :roles, through: :user_roles

  # this has to be after the user_roles association otherwise
  # it gets called twice and causes unique constraint errors
  # https://stackoverflow.com/a/66435239
  after_create :assign_default_role # <----- This works

  def assign_default_role
    contributor_role = Role.find_by!(name: "contributor")
    self.roles << contributor_role unless roles.any?
  end
end
```


It turns out this random solution in stack overflow worked, but I'm new to rails and couldn't find any documentation on ordering of callbacks and _why_ this solution worked

https://stackoverflow.com/questions/40293181/rails-after-create-gets-called-twice/66435239#66435239


```
Your callback after_create :add_user_role must be defined AFTER the associations

# First: 
has_many :directions
has_many :roles, through: :directions

# Then:
after_create :add_user_role

```

**Comment by zacheryph** [comment]
This has to do with how the through: relations work in regard to the creation of the the join records, and the callback ordering.

has_many through: creates callbacks for creating the join records in an after_create callback itself. So you have two things happening here…

Given the example above; When you add a role to a user not yet created, it creates it when the parent record is persisted. If you add a role to an already persisted user, it creates the join record instantly. Now since the user is created when your after callback gets called, it instantly creates the join record. Then the relations callback gets called, and sees that it needs to create the join record itself.

Moving your callback after the relation, puts your callback after the relations, so it sees no roles it has to create join records for.

Now I am not really sure this should be considered a bug. I personally think its really weird to be adding the role in an after callback (no matter where its located.) I’m not sure the relations should be protecting against stuff being created in other callbacks either? Also why should it? The relations after_create callback should be under the assumption that any roles assigned do not have a join record yet (we literally just created the user record.)

A better solution would be to use `before_validate` for this. This ensures everything gets validated properly, say you want to validate user has _at least_ one role. Or you have a role that cannot be assigned to new users, or two roles are mutually exclusive. Get your models how you want them to look & validated _before_ AR even touches the database.
